Handle malloc(0) and malloc(__CPROVER_max_malloc_size) in assigns clause instrumentation

remi-delmas-3000 opened the pull request:
With the previous upper bound expression`ub = lb + size - 1` pointer overflow were possible when the object size was zero.

The expression has been changed to `ub = lb + size` to allow for a zero size without overflows.

The CAR inclusion tests do not need to be changed.

Added tests to demonstrate that both `malloc(0)` and `malloc(__CPROVER_max_malloc_size)` are handled without issue by CAR instrumentation and contract replacement.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

tautschnig commented on regression/contracts/assigns-enforce-malloc-zero/test.desc:
```diff
@@ -0,0 +1,11 @@
+--enforce-contract foo _ --malloc-may-fail
+^EXIT=0$
+^SIGNAL=0$
+^\[foo\.\d+\] line 22 Check that a\[\(signed long int\)i\] is assignable: SUCCESS
```
Use `signed (long (long )?)?int` to ensure this works across several platforms.

remi-delmas-3000 replied:
fixed

tautschnig commented on regression/contracts/assigns-enforce-malloc-zero/test.desc:
```diff
@@ -0,0 +1,11 @@
+^\[foo\.\d+\] line 22 Check that a\[\(signed long int\)i\] is assignable: SUCCESS
+^VERIFICATION SUCCESSFUL$
+--
+This test checks that object of size zero or of __CPROVER_max_malloc_size
```
Nit: "objects" (plural)

remi-delmas-3000 replied:
fixed

tautschnig commented on regression/contracts/assigns-enforce-malloc-zero/main.c:
```diff
@@ -0,0 +1,34 @@
+#include <stdlib.h>
+
+int MAX_SIZE = __CPROVER_max_malloc_size;
```
Nit pick: I'm not even sure why this is necessary, but if it's just about having a shorter name then a `#define` would seem more idiomatic?

jimgrundy replied:
I'm allergic to macros, but if its just about having a shorter name then I'd go with "const int MAX_SIZE = ...;"

remi-delmas-3000 replied:
fixed

tautschnig commented on regression/contracts/assigns-replace-malloc-zero/main.c:
```diff
@@ -0,0 +1,40 @@
+#include <stdlib.h>
+
+int MAX_SIZE = __CPROVER_max_malloc_size;
```
See comment above.

remi-delmas-3000 replied:
fixed

tautschnig commented on src/goto-instrument/contracts/assigns.cpp:
```diff
@@ -133,8 +133,7 @@ assigns_clauset::conditional_address_ranget::generate_snapshot_instructions()
     upper_bound_address_var,
-    minus_exprt{plus_exprt{slice.first, slice.second},
-                from_integer(1, slice.second.type())},
+    plus_exprt{slice.first, slice.second},
```
Why was the "-1" necessary before, or why is this the only change that's necessary?

jimgrundy replied:
I guess we were expecting some "x <= y" checks to become "x < y" checks, so I too am curious to know why that didn't need to happen.

TGWDB replied:
I'd also be curious to know why changing the actual value used does not appear to break anything. Was there a bug before, or is there an edge case now not being tested?

remi-delmas-3000 replied:
these expressions define upper bounds for intervals. We just interpret them as being included/excluded of the interval, but the inclusion check for two intervals just compares the upper bounds using `_ <= _`. Our test suite was missing a case with objects of size 0 and `__CPROVER_max_malloc_size`. My only fear removing the `-1` would be a pointer overflow into the object bits, but it seems like the fact that pointers pointing to one past an object's boundary are legal in C is already factored in `__CPROVER_max_malloc_size`, so everything works.

tautschnig reviewed (commented):
I'm not yet clear why there wasn't any need to change a comparison operator in addition to the single arithmetic change. Could you please clarify?

remi-delmas-3000 commented:
> I'm not yet clear why there wasn't any need to change a comparison operator in addition to the single arithmetic change. Could you please clarify?

these expressions define upper bounds for intervals. We just interpret them as being included/excluded of the interval, but the inclusion check for two intervals just compares the upper bounds using `<=`

feliperodri commented on regression/contracts/assigns-enforce-malloc-zero/main.c:
```diff
@@ -0,0 +1,34 @@
+#include <stdlib.h>
+
+int MAX_SIZE = __CPROVER_max_malloc_size;
+int nondet_int();
```
Why do you need this `nondet_int()`?

feliperodri commented on regression/contracts/assigns-enforce-malloc-zero/main.c:
```diff
@@ -0,0 +1,34 @@
+{
+  if(!a)
+    return -1;
+  int i = nondet_int();
```
If you leave it uninitialized it'll have a nondeterministic value.

remi-delmas-3000 replied:
simplified

feliperodri commented on regression/contracts/assigns-enforce-malloc-zero/test.desc:
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--enforce-contract foo _ --malloc-may-fail
```
You should also include `--malloc-fail-null`.

remi-delmas-3000 replied:
```
 --malloc-may-fail          allow malloc calls to return a null pointer
 --malloc-fail-null           set malloc failure mode to return null
 ```
 `--malloc-may-fail` seems more general than `--malloc-fail-null`, isn't it ?

feliperodri replied:
I'd expect the same thing, but you need the second flag to ensure that `malloc` will return NULL on failure. See the following example.
```c
// file main.c
#include <stdlib.h>

int main()
{
  int *i = malloc(0);
  assert(__CPROVER_DYNAMIC_OBJECT(i));
  //assert(i == NULL);
}
```

If you run it using `cbmc main.c --malloc-may-fail` the result is:
```shell
** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

main.c function main
[main.assertion.1] line 6 assertion IS_DYNAMIC_OBJECT((const void *)i): SUCCESS

** 0 of 3 failed (1 iterations)
VERIFICATION SUCCESSFUL
```

But if you run using `cbmc main.c --malloc-may-fail  --malloc-fail-null --trace`, the result is:
```shell
** Results:
<builtin-library-malloc> function malloc
[malloc.assertion.1] line 26 max allocation size exceeded: SUCCESS
[malloc.assertion.2] line 31 max allocation may fail: SUCCESS

main.c function main
[main.assertion.1] line 6 assertion IS_DYNAMIC_OBJECT((const void *)i): FAILURE

Trace for main.assertion.1:

State 26 file main.c function main line 5 thread 0
----------------------------------------------------
  i=((signed int *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 30 file main.c function main line 5 thread 0
----------------------------------------------------
  malloc_size=0ul (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

State 41 file main.c function main line 5 thread 0
----------------------------------------------------
  i=((signed int *)NULL) (00000000 00000000 00000000 00000000 00000000 00000000 00000000 00000000)

Violated property:
  file main.c function main line 6 thread 0
  assertion IS_DYNAMIC_OBJECT((const void *)i)
  IS_DYNAMIC_OBJECT((const void *)i)



** 1 of 3 failed (2 iterations)
VERIFICATION FAILED
```

remi-delmas-3000 replied:
Fixed. 

Remark: The CLI help and doc should be updated to reflect the actual behaviour.

feliperodri commented on regression/contracts/assigns-replace-malloc-zero/main.c:
```diff
@@ -0,0 +1,40 @@
+#include <stdlib.h>
+
+int MAX_SIZE = __CPROVER_max_malloc_size;
+int nondet_int();
```
Same as above.

remi-delmas-3000 replied:
simplified

feliperodri commented on regression/contracts/assigns-replace-malloc-zero/main.c:
```diff
@@ -0,0 +1,40 @@
+{
+  if(!a)
+    return -1;
+  int i = nondet_int();
```
Same as above.

feliperodri commented on regression/contracts/assigns-replace-malloc-zero/test.desc:
```diff
@@ -0,0 +1,11 @@
+CORE
+main.c
+--replace-call-with-contract foo _ --malloc-may-fail
```
Same as above.

feliperodri commented on regression/contracts/assigns-replace-malloc-zero/test.desc:
```diff
@@ -0,0 +1,11 @@
+\[main\.assertion\.1\] line 38 expecting SUCCESS: SUCCESS$
+^VERIFICATION SUCCESSFUL$
+--
+This test checks that object of size zero or of __CPROVER_max_malloc_size
```
objects

remi-delmas-3000 replied:
fixed

Commit:
Fixed possible pointer overflow in CAR upperbound when the object size is zero.

The expression for the upper bound was ub = lb + size - 1 which could overflow when size == 0.
It is now set to ub = lb + size.

Adding tests that check that CAR instrumentation works with object size of zero
and __CPROVER_max_malloc_size.
